File: examples/example3/example3.cpp

#include "example3.hpp"

namespace example3 {

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

S::~S() noexcept {}
Mod0::~Mod0() noexcept {}
Mod1::~Mod1() noexcept {}
Mod2::~Mod2() noexcept {}
Mod3::~Mod3() noexcept {}
Mod4::~Mod4() noexcept {}
Mod5::~Mod5() noexcept {}
Mod6::~Mod6() noexcept {}


RuleS0::RuleS0( Mod0 const& arg1 )
  :std::tuple< Mod0 >( arg1 ) {}

void RuleS0::accept( Visitor& visitor ) {
  visitor.visitRuleS0( *this );
}
void RuleS0::accept( ConstVisitor& visitor ) const {
  visitor.visitRuleS0( *this );
}


RuleS1::RuleS1( Mod1 const& arg1 )
  :std::tuple< Mod1 >( arg1 ) {}

void RuleS1::accept( Visitor& visitor ) {
  visitor.visitRuleS1( *this );
}
void RuleS1::accept( ConstVisitor& visitor ) const {
  visitor.visitRuleS1( *this );
}


Accept::Accept(  )
  :std::tuple<  >(  ) {}

void Accept::accept( Visitor& visitor ) {
  visitor.visitAccept( *this );
}
void Accept::accept( ConstVisitor& visitor ) const {
  visitor.visitAccept( *this );
}


Rule00::Rule00( Mod0 const& arg1 )
  :std::tuple< Mod0 >( arg1 ) {}

void Rule00::accept( Visitor& visitor ) {
  visitor.visitRule00( *this );
}
void Rule00::accept( ConstVisitor& visitor ) const {
  visitor.visitRule00( *this );
}


Rule01::Rule01( Mod1 const& arg1 )
  :std::tuple< Mod1 >( arg1 ) {}

void Rule01::accept( Visitor& visitor ) {
  visitor.visitRule01( *this );
}
void Rule01::accept( ConstVisitor& visitor ) const {
  visitor.visitRule01( *this );
}


Rule10::Rule10( Mod2 const& arg1 )
  :std::tuple< Mod2 >( arg1 ) {}

void Rule10::accept( Visitor& visitor ) {
  visitor.visitRule10( *this );
}
void Rule10::accept( ConstVisitor& visitor ) const {
  visitor.visitRule10( *this );
}


Rule11::Rule11( Mod3 const& arg1 )
  :std::tuple< Mod3 >( arg1 ) {}

void Rule11::accept( Visitor& visitor ) {
  visitor.visitRule11( *this );
}
void Rule11::accept( ConstVisitor& visitor ) const {
  visitor.visitRule11( *this );
}


Rule20::Rule20( Mod4 const& arg1 )
  :std::tuple< Mod4 >( arg1 ) {}

void Rule20::accept( Visitor& visitor ) {
  visitor.visitRule20( *this );
}
void Rule20::accept( ConstVisitor& visitor ) const {
  visitor.visitRule20( *this );
}


Rule21::Rule21( Mod5 const& arg1 )
  :std::tuple< Mod5 >( arg1 ) {}

void Rule21::accept( Visitor& visitor ) {
  visitor.visitRule21( *this );
}
void Rule21::accept( ConstVisitor& visitor ) const {
  visitor.visitRule21( *this );
}


Rule30::Rule30( Mod6 const& arg1 )
  :std::tuple< Mod6 >( arg1 ) {}

void Rule30::accept( Visitor& visitor ) {
  visitor.visitRule30( *this );
}
void Rule30::accept( ConstVisitor& visitor ) const {
  visitor.visitRule30( *this );
}


Rule31::Rule31( Mod0 const& arg1 )
  :std::tuple< Mod0 >( arg1 ) {}

void Rule31::accept( Visitor& visitor ) {
  visitor.visitRule31( *this );
}
void Rule31::accept( ConstVisitor& visitor ) const {
  visitor.visitRule31( *this );
}


Rule40::Rule40( Mod1 const& arg1 )
  :std::tuple< Mod1 >( arg1 ) {}

void Rule40::accept( Visitor& visitor ) {
  visitor.visitRule40( *this );
}
void Rule40::accept( ConstVisitor& visitor ) const {
  visitor.visitRule40( *this );
}


Rule41::Rule41( Mod2 const& arg1 )
  :std::tuple< Mod2 >( arg1 ) {}

void Rule41::accept( Visitor& visitor ) {
  visitor.visitRule41( *this );
}
void Rule41::accept( ConstVisitor& visitor ) const {
  visitor.visitRule41( *this );
}


Rule50::Rule50( Mod3 const& arg1 )
  :std::tuple< Mod3 >( arg1 ) {}

void Rule50::accept( Visitor& visitor ) {
  visitor.visitRule50( *this );
}
void Rule50::accept( ConstVisitor& visitor ) const {
  visitor.visitRule50( *this );
}


Rule51::Rule51( Mod4 const& arg1 )
  :std::tuple< Mod4 >( arg1 ) {}

void Rule51::accept( Visitor& visitor ) {
  visitor.visitRule51( *this );
}
void Rule51::accept( ConstVisitor& visitor ) const {
  visitor.visitRule51( *this );
}


Rule60::Rule60( Mod5 const& arg1 )
  :std::tuple< Mod5 >( arg1 ) {}

void Rule60::accept( Visitor& visitor ) {
  visitor.visitRule60( *this );
}
void Rule60::accept( ConstVisitor& visitor ) const {
  visitor.visitRule60( *this );
}


Rule61::Rule61( Mod6 const& arg1 )
  :std::tuple< Mod6 >( arg1 ) {}

void Rule61::accept( Visitor& visitor ) {
  visitor.visitRule61( *this );
}
void Rule61::accept( ConstVisitor& visitor ) const {
  visitor.visitRule61( *this );
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

std::ostream& operator <<( std::ostream &out, S const& self ) {
  class Visitor : public S::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitRuleS0( RuleS0 const& host ) {
      *out_ << host;
    }
    void visitRuleS1( RuleS1 const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}

std::ostream& operator <<( std::ostream &out, Mod0 const& self ) {
  class Visitor : public Mod0::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitAccept( Accept const& host ) {
      *out_ << host;
    }
    void visitRule00( Rule00 const& host ) {
      *out_ << host;
    }
    void visitRule01( Rule01 const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}

std::ostream& operator <<( std::ostream &out, Mod1 const& self ) {
  class Visitor : public Mod1::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitRule10( Rule10 const& host ) {
      *out_ << host;
    }
    void visitRule11( Rule11 const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}

std::ostream& operator <<( std::ostream &out, Mod2 const& self ) {
  class Visitor : public Mod2::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitRule20( Rule20 const& host ) {
      *out_ << host;
    }
    void visitRule21( Rule21 const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}

std::ostream& operator <<( std::ostream &out, Mod3 const& self ) {
  class Visitor : public Mod3::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitRule30( Rule30 const& host ) {
      *out_ << host;
    }
    void visitRule31( Rule31 const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}

std::ostream& operator <<( std::ostream &out, Mod4 const& self ) {
  class Visitor : public Mod4::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitRule40( Rule40 const& host ) {
      *out_ << host;
    }
    void visitRule41( Rule41 const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}

std::ostream& operator <<( std::ostream &out, Mod5 const& self ) {
  class Visitor : public Mod5::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitRule50( Rule50 const& host ) {
      *out_ << host;
    }
    void visitRule51( Rule51 const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}

std::ostream& operator <<( std::ostream &out, Mod6 const& self ) {
  class Visitor : public Mod6::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitRule60( Rule60 const& host ) {
      *out_ << host;
    }
    void visitRule61( Rule61 const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}


std::ostream& operator <<( std::ostream& out, RuleS0 const& self ) {
  out << "RuleS0("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, RuleS1 const& self ) {
  out << "RuleS1("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Accept const& self ) {
  out << "Accept(" << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule00 const& self ) {
  out << "Rule00("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule01 const& self ) {
  out << "Rule01("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule10 const& self ) {
  out << "Rule10("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule11 const& self ) {
  out << "Rule11("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule20 const& self ) {
  out << "Rule20("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule21 const& self ) {
  out << "Rule21("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule30 const& self ) {
  out << "Rule30("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule31 const& self ) {
  out << "Rule31("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule40 const& self ) {
  out << "Rule40("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule41 const& self ) {
  out << "Rule41("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule50 const& self ) {
  out << "Rule50("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule51 const& self ) {
  out << "Rule51("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule60 const& self ) {
  out << "Rule60("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Rule61 const& self ) {
  out << "Rule61("<<  std::get< 0 >( self ) << ")";
  return out;
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

Node1::Node1() {}

Node2::Node2( S const& content_ ) :content( content_ ) {}

Node3::Node3() {}

Node4::Node4() {}

Node5::Node5() {}

Node6::Node6( Mod0 const& content_ ) :content( content_ ) {}

Node7::Node7( Mod1 const& content_ ) :content( content_ ) {}

Node8::Node8() {}

Node9::Node9( Mod2 const& content_ ) :content( content_ ) {}

Node10::Node10() {}

Node11::Node11() {}

Node12::Node12() {}

Node13::Node13( Mod3 const& content_ ) :content( content_ ) {}

Node14::Node14() {}

Node15::Node15( Mod4 const& content_ ) :content( content_ ) {}

Node16::Node16() {}

Node17::Node17() {}

Node18::Node18( Mod5 const& content_ ) :content( content_ ) {}

Node19::Node19() {}

Node20::Node20( Mod6 const& content_ ) :content( content_ ) {}

Node21::Node21() {}

Node22::Node22() {}

Node23::Node23( Mod0 const& content_ ) :content( content_ ) {}

Node24::Node24( Mod1 const& content_ ) :content( content_ ) {}

Node25::Node25() {}

Node26::Node26( Mod2 const& content_ ) :content( content_ ) {}

Node27::Node27( Mod3 const& content_ ) :content( content_ ) {}

Node28::Node28() {}

Node29::Node29( Mod4 const& content_ ) :content( content_ ) {}

Node30::Node30( Mod5 const& content_ ) :content( content_ ) {}

Node31::Node31() {}

Node32::Node32( Mod6 const& content_ ) :content( content_ ) {}

Node33::Node33( Mod0 const& content_ ) :content( content_ ) {}

Node34::Node34( Mod1 const& content_ ) :content( content_ ) {}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

std::shared_ptr< State< Node1 > > begin() {
  std::shared_ptr< State<> > bottom( new State<>() );
  return State< Node1 >::make( Node1(), bottom );
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

}

File: examples/example3/Example3.hs

{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE UndecidableInstances #-}

module Example3 where

-------------------------------------------------------------------------------

-- grammar definition

-- ruleS0 : S -> "O()" Mod0
-- ruleS1 : S -> "I()" Mod1
-- accept : Mod0 -> eps
-- rule00 : Mod0 -> "O()" Mod0
-- rule01 : Mod0 -> "I()" Mod1
-- rule10 : Mod1 -> "O()" Mod2
-- rule11 : Mod1 -> "I()" Mod3
-- rule20 : Mod2 -> "O()" Mod4
-- rule21 : Mod2 -> "I()" Mod5
-- rule30 : Mod3 -> "O()" Mod6
-- rule31 : Mod3 -> "I()" Mod0
-- rule40 : Mod4 -> "O()" Mod1
-- rule41 : Mod4 -> "I()" Mod2
-- rule50 : Mod5 -> "O()" Mod3
-- rule51 : Mod5 -> "I()" Mod4
-- rule60 : Mod6 -> "O()" Mod5
-- rule61 : Mod6 -> "I()" Mod6

-------------------------------------------------------------------------------

type Program r a = (a -> r) -> r

program :: a -> Program r a
program a = \k -> k a

-------------------------------------------------------------------------------

-- AST nodes

data S
  = RuleS0 Mod0
  | RuleS1 Mod1
  deriving (Show)

data Mod0
  = Accept
  | Rule00 Mod0
  | Rule01 Mod1
  deriving (Show)

data Mod1
  = Rule10 Mod2
  | Rule11 Mod3
  deriving (Show)

data Mod2
  = Rule20 Mod4
  | Rule21 Mod5
  deriving (Show)

data Mod3
  = Rule30 Mod6
  | Rule31 Mod0
  deriving (Show)

data Mod4
  = Rule40 Mod1
  | Rule41 Mod2
  deriving (Show)

data Mod5
  = Rule50 Mod3
  | Rule51 Mod4
  deriving (Show)

data Mod6
  = Rule60 Mod5
  | Rule61 Mod6
  deriving (Show)

-------------------------------------------------------------------------------

-- terminal symbols

class OTransition s t | s -> t where
  o :: s -> Program r t

class ITransition s t | s -> t where
  i :: s -> Program r t

-------------------------------------------------------------------------------

-- automaton states

data Node1 prev = Node1 prev

data Node2 prev = Node2 prev S

data Node3 prev = Node3 prev

data Node4 prev = Node4 prev

data Node5 prev = Node5 prev

data Node6 prev = Node6 prev Mod0

data Node7 prev = Node7 prev Mod1

data Node8 prev = Node8 prev

data Node9 prev = Node9 prev Mod2

data Node10 prev = Node10 prev

data Node11 prev = Node11 prev

data Node12 prev = Node12 prev

data Node13 prev = Node13 prev Mod3

data Node14 prev = Node14 prev

data Node15 prev = Node15 prev Mod4

data Node16 prev = Node16 prev

data Node17 prev = Node17 prev

data Node18 prev = Node18 prev Mod5

data Node19 prev = Node19 prev

data Node20 prev = Node20 prev Mod6

data Node21 prev = Node21 prev

data Node22 prev = Node22 prev

data Node23 prev = Node23 prev Mod0

data Node24 prev = Node24 prev Mod1

data Node25 prev = Node25 prev

data Node26 prev = Node26 prev Mod2

data Node27 prev = Node27 prev Mod3

data Node28 prev = Node28 prev

data Node29 prev = Node29 prev Mod4

data Node30 prev = Node30 prev Mod5

data Node31 prev = Node31 prev

data Node32 prev = Node32 prev Mod6

data Node33 prev = Node33 prev Mod0

data Node34 prev = Node34 prev Mod1

-------------------------------------------------------------------------------

-- shift transitions

instance {-# OVERLAPS #-} ITransition (Node1 prev) (Node12 (Node1 prev)) where
  i prev = program (Node12 prev)

instance {-# OVERLAPS #-} ITransition (Node3 prev) (Node8 (Node3 prev)) where
  i prev = program (Node8 prev)

instance {-# OVERLAPS #-} ITransition (Node4 prev) (Node8 (Node4 prev)) where
  i prev = program (Node8 prev)

instance {-# OVERLAPS #-} ITransition (Node5 prev) (Node8 (Node5 prev)) where
  i prev = program (Node8 prev)

instance {-# OVERLAPS #-} ITransition (Node8 prev) (Node14 (Node8 prev)) where
  i prev = program (Node14 prev)

instance {-# OVERLAPS #-} ITransition (Node10 prev) (Node19 (Node10 prev)) where
  i prev = program (Node19 prev)

instance {-# OVERLAPS #-} ITransition (Node11 prev) (Node14 (Node11 prev)) where
  i prev = program (Node14 prev)

instance {-# OVERLAPS #-} ITransition (Node12 prev) (Node14 (Node12 prev)) where
  i prev = program (Node14 prev)

instance {-# OVERLAPS #-} ITransition (Node14 prev) (Node4 (Node14 prev)) where
  i prev = program (Node4 prev)

instance {-# OVERLAPS #-} ITransition (Node16 prev) (Node17 (Node16 prev)) where
  i prev = program (Node17 prev)

instance {-# OVERLAPS #-} ITransition (Node17 prev) (Node19 (Node17 prev)) where
  i prev = program (Node19 prev)

instance {-# OVERLAPS #-} ITransition (Node19 prev) (Node25 (Node19 prev)) where
  i prev = program (Node25 prev)

instance {-# OVERLAPS #-} ITransition (Node21 prev) (Node31 (Node21 prev)) where
  i prev = program (Node31 prev)

instance {-# OVERLAPS #-} ITransition (Node22 prev) (Node4 (Node22 prev)) where
  i prev = program (Node4 prev)

instance {-# OVERLAPS #-} ITransition (Node25 prev) (Node17 (Node25 prev)) where
  i prev = program (Node17 prev)

instance {-# OVERLAPS #-} ITransition (Node28 prev) (Node25 (Node28 prev)) where
  i prev = program (Node25 prev)

instance {-# OVERLAPS #-} ITransition (Node31 prev) (Node31 (Node31 prev)) where
  i prev = program (Node31 prev)


instance {-# OVERLAPS #-} OTransition (Node1 prev) (Node5 (Node1 prev)) where
  o prev = program (Node5 prev)

instance {-# OVERLAPS #-} OTransition (Node3 prev) (Node3 (Node3 prev)) where
  o prev = program (Node3 prev)

instance {-# OVERLAPS #-} OTransition (Node4 prev) (Node3 (Node4 prev)) where
  o prev = program (Node3 prev)

instance {-# OVERLAPS #-} OTransition (Node5 prev) (Node3 (Node5 prev)) where
  o prev = program (Node3 prev)

instance {-# OVERLAPS #-} OTransition (Node8 prev) (Node10 (Node8 prev)) where
  o prev = program (Node10 prev)

instance {-# OVERLAPS #-} OTransition (Node10 prev) (Node16 (Node10 prev)) where
  o prev = program (Node16 prev)

instance {-# OVERLAPS #-} OTransition (Node11 prev) (Node10 (Node11 prev)) where
  o prev = program (Node10 prev)

instance {-# OVERLAPS #-} OTransition (Node12 prev) (Node10 (Node12 prev)) where
  o prev = program (Node10 prev)

instance {-# OVERLAPS #-} OTransition (Node14 prev) (Node21 (Node14 prev)) where
  o prev = program (Node21 prev)

instance {-# OVERLAPS #-} OTransition (Node16 prev) (Node11 (Node16 prev)) where
  o prev = program (Node11 prev)

instance {-# OVERLAPS #-} OTransition (Node17 prev) (Node16 (Node17 prev)) where
  o prev = program (Node16 prev)

instance {-# OVERLAPS #-} OTransition (Node19 prev) (Node22 (Node19 prev)) where
  o prev = program (Node22 prev)

instance {-# OVERLAPS #-} OTransition (Node21 prev) (Node28 (Node21 prev)) where
  o prev = program (Node28 prev)

instance {-# OVERLAPS #-} OTransition (Node22 prev) (Node21 (Node22 prev)) where
  o prev = program (Node21 prev)

instance {-# OVERLAPS #-} OTransition (Node25 prev) (Node11 (Node25 prev)) where
  o prev = program (Node11 prev)

instance {-# OVERLAPS #-} OTransition (Node28 prev) (Node22 (Node28 prev)) where
  o prev = program (Node22 prev)

instance {-# OVERLAPS #-} OTransition (Node31 prev) (Node28 (Node31 prev)) where
  o prev = program (Node28 prev)

-------------------------------------------------------------------------------

-- reduces

class Reduce s t | s -> t where
  reduce :: s -> t


-- ruleS0 : S -> "O()" Mod0

instance Reduce (Node33 (Node5 (Node1 prev))) (Node2 (Node1 prev)) where
  reduce (Node33 (Node5 prev) x1) = (Node2 prev (RuleS0 x1))


-- ruleS1 : S -> "I()" Mod1

instance Reduce (Node34 (Node12 (Node1 prev))) (Node2 (Node1 prev)) where
  reduce (Node34 (Node12 prev) x1) = (Node2 prev (RuleS1 x1))


-- accept : Mod0 -> eps

instance Reduce (Node3 prev) (Node6 (Node3 prev)) where
  reduce prev = (Node6 prev (Accept))

instance Reduce (Node4 prev) (Node23 (Node4 prev)) where
  reduce prev = (Node23 prev (Accept))

instance Reduce (Node5 prev) (Node33 (Node5 prev)) where
  reduce prev = (Node33 prev (Accept))


-- rule00 : Mod0 -> "O()" Mod0

instance Reduce (Node6 (Node3 (Node3 prev))) (Node6 (Node3 prev)) where
  reduce (Node6 (Node3 prev) x1) = (Node6 prev (Rule00 x1))

instance Reduce (Node6 (Node3 (Node4 prev))) (Node23 (Node4 prev)) where
  reduce (Node6 (Node3 prev) x1) = (Node23 prev (Rule00 x1))

instance Reduce (Node6 (Node3 (Node5 prev))) (Node33 (Node5 prev)) where
  reduce (Node6 (Node3 prev) x1) = (Node33 prev (Rule00 x1))


-- rule01 : Mod0 -> "I()" Mod1

instance Reduce (Node7 (Node8 (Node3 prev))) (Node6 (Node3 prev)) where
  reduce (Node7 (Node8 prev) x1) = (Node6 prev (Rule01 x1))

instance Reduce (Node7 (Node8 (Node4 prev))) (Node23 (Node4 prev)) where
  reduce (Node7 (Node8 prev) x1) = (Node23 prev (Rule01 x1))

instance Reduce (Node7 (Node8 (Node5 prev))) (Node33 (Node5 prev)) where
  reduce (Node7 (Node8 prev) x1) = (Node33 prev (Rule01 x1))


-- rule10 : Mod1 -> "O()" Mod2

instance Reduce (Node9 (Node10 (Node8 prev))) (Node7 (Node8 prev)) where
  reduce (Node9 (Node10 prev) x1) = (Node7 prev (Rule10 x1))

instance Reduce (Node9 (Node10 (Node11 prev))) (Node24 (Node11 prev)) where
  reduce (Node9 (Node10 prev) x1) = (Node24 prev (Rule10 x1))

instance Reduce (Node9 (Node10 (Node12 prev))) (Node34 (Node12 prev)) where
  reduce (Node9 (Node10 prev) x1) = (Node34 prev (Rule10 x1))


-- rule11 : Mod1 -> "I()" Mod3

instance Reduce (Node13 (Node14 (Node8 prev))) (Node7 (Node8 prev)) where
  reduce (Node13 (Node14 prev) x1) = (Node7 prev (Rule11 x1))

instance Reduce (Node13 (Node14 (Node11 prev))) (Node24 (Node11 prev)) where
  reduce (Node13 (Node14 prev) x1) = (Node24 prev (Rule11 x1))

instance Reduce (Node13 (Node14 (Node12 prev))) (Node34 (Node12 prev)) where
  reduce (Node13 (Node14 prev) x1) = (Node34 prev (Rule11 x1))


-- rule20 : Mod2 -> "O()" Mod4

instance Reduce (Node15 (Node16 (Node10 prev))) (Node9 (Node10 prev)) where
  reduce (Node15 (Node16 prev) x1) = (Node9 prev (Rule20 x1))

instance Reduce (Node15 (Node16 (Node17 prev))) (Node26 (Node17 prev)) where
  reduce (Node15 (Node16 prev) x1) = (Node26 prev (Rule20 x1))


-- rule21 : Mod2 -> "I()" Mod5

instance Reduce (Node18 (Node19 (Node10 prev))) (Node9 (Node10 prev)) where
  reduce (Node18 (Node19 prev) x1) = (Node9 prev (Rule21 x1))

instance Reduce (Node18 (Node19 (Node17 prev))) (Node26 (Node17 prev)) where
  reduce (Node18 (Node19 prev) x1) = (Node26 prev (Rule21 x1))


-- rule30 : Mod3 -> "O()" Mod6

instance Reduce (Node20 (Node21 (Node14 prev))) (Node13 (Node14 prev)) where
  reduce (Node20 (Node21 prev) x1) = (Node13 prev (Rule30 x1))

instance Reduce (Node20 (Node21 (Node22 prev))) (Node27 (Node22 prev)) where
  reduce (Node20 (Node21 prev) x1) = (Node27 prev (Rule30 x1))


-- rule31 : Mod3 -> "I()" Mod0

instance Reduce (Node23 (Node4 (Node14 prev))) (Node13 (Node14 prev)) where
  reduce (Node23 (Node4 prev) x1) = (Node13 prev (Rule31 x1))

instance Reduce (Node23 (Node4 (Node22 prev))) (Node27 (Node22 prev)) where
  reduce (Node23 (Node4 prev) x1) = (Node27 prev (Rule31 x1))


-- rule40 : Mod4 -> "O()" Mod1

instance Reduce (Node24 (Node11 (Node16 prev))) (Node15 (Node16 prev)) where
  reduce (Node24 (Node11 prev) x1) = (Node15 prev (Rule40 x1))

instance Reduce (Node24 (Node11 (Node25 prev))) (Node29 (Node25 prev)) where
  reduce (Node24 (Node11 prev) x1) = (Node29 prev (Rule40 x1))


-- rule41 : Mod4 -> "I()" Mod2

instance Reduce (Node26 (Node17 (Node16 prev))) (Node15 (Node16 prev)) where
  reduce (Node26 (Node17 prev) x1) = (Node15 prev (Rule41 x1))

instance Reduce (Node26 (Node17 (Node25 prev))) (Node29 (Node25 prev)) where
  reduce (Node26 (Node17 prev) x1) = (Node29 prev (Rule41 x1))


-- rule50 : Mod5 -> "O()" Mod3

instance Reduce (Node27 (Node22 (Node19 prev))) (Node18 (Node19 prev)) where
  reduce (Node27 (Node22 prev) x1) = (Node18 prev (Rule50 x1))

instance Reduce (Node27 (Node22 (Node28 prev))) (Node30 (Node28 prev)) where
  reduce (Node27 (Node22 prev) x1) = (Node30 prev (Rule50 x1))


-- rule51 : Mod5 -> "I()" Mod4

instance Reduce (Node29 (Node25 (Node19 prev))) (Node18 (Node19 prev)) where
  reduce (Node29 (Node25 prev) x1) = (Node18 prev (Rule51 x1))

instance Reduce (Node29 (Node25 (Node28 prev))) (Node30 (Node28 prev)) where
  reduce (Node29 (Node25 prev) x1) = (Node30 prev (Rule51 x1))


-- rule60 : Mod6 -> "O()" Mod5

instance Reduce (Node30 (Node28 (Node21 prev))) (Node20 (Node21 prev)) where
  reduce (Node30 (Node28 prev) x1) = (Node20 prev (Rule60 x1))

instance Reduce (Node30 (Node28 (Node31 prev))) (Node32 (Node31 prev)) where
  reduce (Node30 (Node28 prev) x1) = (Node32 prev (Rule60 x1))


-- rule61 : Mod6 -> "I()" Mod6

instance Reduce (Node32 (Node31 (Node21 prev))) (Node20 (Node21 prev)) where
  reduce (Node32 (Node31 prev) x1) = (Node20 prev (Rule61 x1))

instance Reduce (Node32 (Node31 (Node31 prev))) (Node32 (Node31 prev)) where
  reduce (Node32 (Node31 prev) x1) = (Node32 prev (Rule61 x1))


-- Reduce -> Transition

instance {-# OVERLAPS #-} (Reduce r s, OTransition s t) => OTransition r t where
  o r = o (reduce r)

instance {-# OVERLAPS #-} (Reduce r s, ITransition s t) => ITransition r t where
  i r = i (reduce r)

-------------------------------------------------------------------------------

-- initial state

begin :: Program r (Node1 ())
begin = program (Node1 ())

-------------------------------------------------------------------------------

class End s r | s -> r where
  end :: s -> r

instance {-# OVERLAPS #-} End (Node2 prev) S where
  end (Node2 _ r) = r

instance {-# OVERLAPS #-} (Reduce r s, End s t) => End r t where
  end r = end (reduce r)

-------------------------------------------------------------------------------

